### ISSUE TITLE
Change config functionality

### ISSUE THREAD

**Issue by nicksrandall** [kotojs/kotojs#9]
## TL;DR

I'm thinking about changing how configs work in KotoJS. The basic functionality will remain the same but the advanced functionality would be replaced in favor of a better alternative. 
## My plan

The `config` method will remain very similar to d3's `attr` and `style` methods as it will be a basic getter/setter for config properties. 

``` js
chart.config('height', 500); // sets height config to 500
chart.config('height'); // => returns config value (500)
```

When defining the default config options for a chart, a user will be able to optionally set a few basic validation and meta-data like properties on the config object to provide consumers of the chart with support and documentation of what configs are available on a specific chart as well as what type(s) of values it is looking for. This idea is inspired by [Joi.js](https://github.com/hapijs/joi), an excellent validation library. We probably won't use Joi directly because it's too heavy for this use case. I'm thinking that I will build a "lighter" version of this lib that internally uses some of d3's methods for validation.

``` js
import C from 'koto-config';

// set in the chart somewhere
this._config  = {
  height: C.number().positivie().default(500).decription('the height of the chart').units('px'),
  width: C.number().positivie().default(500).decription('the height of the chart').units('px'),
  fill: C.color(),
  orientation: C.string().in(['vertical', 'horizontal']).default('vertical'),
  xScale: C.scale()
};

// later when user tries to set a config
chart.config('height', -100); 
  // => throws error with helpful message like:
  // "the value for the `height` config for `koto-bar-chart`, decribed as "the height of the chart", was expecting a positive integer (in "px") and got -100. 
```

Again, this would all be optional. Authors of Koto charts would still be able to setup default config values like so and the only validation that would occur when getting or setting a config value is that that particular config exists on the chart. 

``` js
this._config  = {
  height: 500,
  width: 500,
  fill: '#9ce',
  orientation: 'vertical',
  xScale: d3.scale.linear()
};
```

Before I make this change, I want to get some validation/feedback on the idea. If you don't like it or have a better idea, please let me know!

**Comment by girlwithglasses** [comment]
I hadn't seen joi before -- it looks interesting! The schema that it validates against looks similar to JSON schema, which has the advantage of being able to use a single schema for browser and server validation.

The potential problem I see is that there may need to be a large set of functions used for validation -- e.g. `colour()`, `positive()`, `string()`, etc. A possible fix for this would be to have the config validator as a mixin (rather than built into the chart, as the current `config` functionality is), so users could choose whether or not they want to use it.

I am currently working on a variation on your `config` method for generating standard getters and setters for each config parameter, and to specify a custom getter/setter if needed. At the moment I am not doing any validation but it's easy to imagine it being specified as part of the setter (or autogenerated from the config). I adapted a pattern from [this article](http://raganwald.com/2015/08/24/ready-get-set-go.html) to build a function that will take a JSON schema-type data structure and generate attributes for the class from it. Here's a simplified version:

```
// default setter: sets the attribute, triggers a change event
var setter = function( ns, key, value ){
	// could add in validation here
	var attr = ns ? this[ns].attr : this.attr;
	attr.set( key, value );
	this.trigger( 'change:' + ( ns ? ns + ':' : '' ) + key );
}
// default getter: gets the attribute
, getter = function( ns, key ){
	var attr = ns ? this[ns].attr : this.attr;
	if ( attr ) {
		return attr.get( key );
	}
	return;
};

function add_attributes( target, props, ns = null ) {

	var attr_target = target;

	// store all the attribute data in a Map
	if ('undefined' === typeof attr_target.attr) {
		attr_target.attr = new Map();
	}

	for (let propName in props) {
		let prop = props[propName];
		// read-only property
		if ( prop.default && prop.read_only ) {
			Object.defineProperty( attr_target, propName, {
				value: prop.default,
				enumerable: true
			});
		}
		else {
			Object.defineProperty( attr_target, propName, {
				get: prop.getter || getter.bind( target, ns, propName ),
				set: prop.setter || setter.bind( target, ns, propName ),
				configurable: true,
				enumerable: true
			});
			if ( prop.default ) {
				attr_target[propName] = prop.default;
			}
		}
	}
	return target;
}
```

It is pretty similar to what you had written in a previous version of Koto. There is also a `set_attr` method that enables chaining by running the setter and returning the `chart` object.

An example of a config:

```
var init_cfg = {
	colorbrewer: {
		type: 'string'
		, setter: function(_) {
			this.base.classed( this.colorbrewer, false );
			this.attr.set('colorbrewer', _);
			this.trigger('change:colors');
			this.base.classed( this.colorbrewer, true );
		}
		, default: 'Blues'
	}
	, duration: {
		type: 'number'
		, default: 750
		, min: 250
		, max: 5000

	}
	, unique_id: {
		type: 'string'
		, default: '_id'
	}
	, schema: {
		type: 'object'
	}
};
```

So I can do e.g. `chart.colorbrewer = 'Reds'`, and it runs the setter code, which includes triggering a chart event and setting the class of the chart base.

I have several different charts and I also wanted to abstract out common functionality into mixins that can be applied to any chart; e.g. sorting; dimensions [height, width, margins]; axes; labels; tooltip; etc. Each mixin can have its own configuration that gets integrated into the configuration for the whole chart. I had been using a fairly simple mixin function that added sugar to the existing `chart X extends Y` syntax, but ran into problems with initialisation order and merging configs. I hence migrated to the [universal mixin](https://github.com/WebReflection/universal-mixin) for the mixins, and applying them as ES7 decorators like so:

```
// a mixin for sorting data
var sortable_mixin = mixin({

	get sort_cfg() {
		return {
		// data structure representing the attributes for the mixin
			sort: {
				type: 'object',
				properties: {
					dimension: {
						type: 'string or function',
						description: 'a function or string representing the aspect of the object to be sorted, e.g. an object key'
					},
					order: {
						type: 'enum',
						description: 'the sort to apply to the objects, e.g. num_asc for an ascending numerical sort; _DESC_ for a descending case-insensitive natural sort',
						enum: [
							'_A_Z_', '_Z_A_', '_a_z_', '_z_a_', 'num_asc', 'num_desc'
						]
					},
					custom: {
						type: 'function',
						description: 'a custom sort function'
					}
				}
			}
			, sortable: {
				type: 'boolean',
				setter: function( args ){
					// snip long setter function -- sets the sort.dimension and sort.order above
				}
			}
		};
	}
	, init() {
		this.proto_config.push( this.sort_cfg );
		this.on('change:sortable', () => {
			if ( this.data ) {
				this.data = this._sort( this.sortable, this.data );
				this.draw( this.data );
			}
		});
	}

	, _sort( data ) {
		// the actual data sorting function
	}
});
```

```
@dimension_mixin
@axis_mixin
class BarChart extends Koto {
	constructor( selection, options = {} ) {
		super( selection, options );
		// configuration for this chart
		this.proto_config = [ init_cfg ];
		// I added an 'initialise' hook that is called before the first chart.draw
		this.on('initialise', function(){
			this.init();
		});
	}

	init(){
		// the chart and mixin configs are all in the array this.proto_config
		// use function add_attributes (above) to create the chart config
		var cfg = this.merge_arr( this.proto_config ); // merge into a single object
		add_attributes( this, cfg );
		// other chart initialisation stuff goes here -- e.g. creating scales,
		// setting up variables, etc., etc.
	}
...
}

@dimension_mixin
@sortable_mixin
class HeatMap {
...
}
```

Applying the universal mixin to a class and giving the class an `init` method will call `init` in all the mixins; I use that to merge the class config with the mixin configs.

I was part-way through migrating my current charts to this model (which mostly involves shifting the chart set up from the `constructor` to a different function) and making some decisions on how to do the config merging and attribute set up when I had to stop to work on something else, so the implementation is incomplete at present.

Let me know if that makes sense!

**Comment by aharris88** [comment]
That's really interesting. I had never heard of Joi, but I like this idea a lot. The one I really like is the `in` validation. My question is, where are these functions defined? You gave some examples:

    number
    positive
    default
    description
    units
    color
    string
    in
    scale

Would these be defined in koto, or in the individual charts? or a combination of both?

**Comment by nicksrandall** [comment]
@aharris88 They would be defined in a new, optional module that you could choose to include with Koto.

**Comment by aharris88** [comment]
And what if the optional module didn't have everything that you needed. I assume you could add your own module or functions?

**Comment by aharris88** [comment]
@nicksrandall I like that idea. So you could do something like this:

    import C from 'koto-config';

    // set in the chart somewhere
    this._config  = {
      height: C.custom(even, 'Must be even')
    };

    function even(_) {
      return !(_ % 2);
    }

But I think there might be a better way to specify the error message.

**Comment by nicksrandall** [comment]
@aharris88 @girlwithglasses,

It's taken me longer than I would have liked to spend the necessary time on implementing this but I'm happy to report that I have a work in progress and that my idea has changed slightly since my last post.

After reviewing a few of the options, I have decided that I like @girlwithglasses's idea of using [JSON-Schema](http://json-schema.org/) syntax (an open standard) to write the validation constraints. I have found a great (and fast) library called [jjv](https://github.com/acornejo/jjv) that we can leverage for our basic type checking and it allows users to write "custom" types that we could leverage for things like "colors".

I would like to keep this extended functionality optional so I'm thinking it should live in a separate repo within the Koto organization. 

In order to use this extended functionality, I'm thinking that a decorator would work really well (after all, this is a decorator's exact use-case). Here is what that would look like: 

```js
// JSON SCHEMA
const schema = {
  type: 'object',
  properties: {
    height: {
      type: 'number',
      minimum: 500
    }
  }
};

// some koto chart
@KotoConfig(schema)
class Chart extends Koto {
  constructor(selection) {
    super(selection);
    this.configs = {
      height: 500
    };
  }
  preDraw() {
    // do something on predraw
  }
}

const chart = new Chart(d3.select(document.body));
chart.config('height', 400); // => throws Error
```

I'll also point out that implementing this will introduce a breaking change. The configs will no longer need to be a "definition" object but it will simply be key-value pairs.

Again, I'm not done implementing this and everything is subject to change. I would love to get your feedback as I move forward. Would this work for your use-cases of KotoJS?

Feel free to look at the code I'm working on here: https://github.com/kotojs/koto-config
